src/AWSLambda/Handler.hs

```diff
@@ -79,25 +85,49 @@ lambdaMain ::
   => (event -> IO res) -- ^ Function to process the event
   -> IO ()
 lambdaMain act =
-  withResultChannel $ \resultChannel -> do
-    input <- ByteString.getLine
-    case Aeson.eitherDecodeStrict input of
-      Left err -> error err
-      Right event -> do
-        result <- act event
-        Text.hPutStrLn resultChannel $ Aeson.encodeToLazyText result
-        pure ()
-
--- | Invoke an action with the handle to write the results to. On AWS Lambda,
--- use the channel opened by the JavaScript wrapper, otherwise use standard
--- output.
-withResultChannel :: (Handle -> IO r) -> IO r
-withResultChannel act = do
-  lambdaFunctionName <- lookupEnv "AWS_LAMBDA_FUNCTION_NAME"
-  case lambdaFunctionName of
-    Just _ -> bracket (fdToHandle communicationFd) hClose act
-    Nothing -> act stdout
-
--- | File descriptor opened by the JavaScript wrapper to listen for the results
-communicationFd :: Fd
-communicationFd = Fd 3
+  isLambda >>= \case
+    True -> lambdaMain' act
+    False -> localMain act
+
+-- | Process the requests from the JavaScript wrapper using the provided action.
+lambdaMain' ::
+     (Aeson.FromJSON event, Aeson.ToJSON res)
+  => (event -> IO res) -- ^ Function to process the event
+  -> IO ()
+lambdaMain' act = do
+  listenSocket <- listenOn communicationPort
+  forever $ do
+    (socket, _, _) <- accept listenSocket
+    forkIO $ finally (hClose socket) $ handleEvent socket socket act
```

Comment:
I was thinking about this last night and I'm concerned about how this handles errors. If `act` throws an exception how does anyone know about it? If I understand the code correctly, the Node wrapper will only report the error if it can't parse valid JSON from the output. But it's possible that the output could contain valid JSON and still throw an exception. I think it would be better if an exception thrown from `act` either terminated the Haskell program with an error status, or sent an error result back over the socket so that the Node wrapper could properly report it.

Reply:
How can anything be written to the socket if an exception is thrown? There's no value to convert and write, so the socket will be closed with no output.

Reply:
What if `act` writes some valid JSON to the socket and then throws an exception? Or do you think it's ok to ignore anything that happens after the output is sent?

Reply:
`act` doesn't have access to a socket. Nothing is written to the socket before it completes. The only way I can see to raise an exception after anything is written to the socket is return something like `[undefined]` from `act` - but I think that'll always result in invalid JSON.

In other words, the only situation valid JSON is written to the output is when https://github.com/seek-oss/serverless-haskell/pull/28/files#diff-d268daad27c8c2f18c333c2ee23bdbf3R125 completes successfully.

I wonder if this was proven already somewhere, but I also think no valid JSON is a token-level prefix of other valid JSON ("token-level" added to disallow the trivial case of `1` and `11`).